feat: Add recipe as ingredient

parumpum opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.

  To start, try providing a short summary of your changes in the Title above. We follow Conventional Commits syntax, please ensure your title is prefixed with one of:
  - `feat: `
  - `fix: `
  - `docs: `
  - `chore: `
  - `dev:`

  If a section of the PR template does not apply to this PR, then delete that section.

  PLEASE READ:
  -------------------------
  Mealie is moving to a regular, automatic release schedule. This means that all PRs should be in a
  stable state, ready for release. This includes:

  - Ensuring new tests have been added to cover new features, or to prevent regressions.
  - Work is fully complete and usable

 -->

## What type of PR is this?

- feature

## What this PR does / why we need it:

Adds the ability to reference another recipe in a recipe's ingredients. A reference to the recipe is added to a new column for table recipes_ingredients. This can be utilized to determine if the ingredient is a food, a recipe, or just a note.

Adding recipe to shopping list will loop through the ingredients list, cascading down into recipes if it finds any to add individual ingredients to the shopping list.

<img width="1307" height="239" alt="image" src="https://github.com/user-attachments/assets/57379833-c2d2-41b1-9f34-e2751a4e724e" />



<img width="399" alt="image" src="https://github.com/user-attachments/assets/c63d84bc-9604-4dfc-b3da-494c4a1b6e78" />

Clicking the referenced recipe will open a new tab to the recipe

Linked recipes will expand into ingredients to add to shopping lists.

## Which issue(s) this PR fixes:

https://github.com/mealie-recipes/mealie/discussions/1761

## Testing

Tested by linking recipes together, adding ingredients to shopping lists
Unit test that adds a recipe to another recipe as an ingredient
Modified referenced recipe name and ingredients, verified changes were picked up on main recipe page

parumpum committed:
Initial recipe linking changes

parumpum committed:
Add db migration to create recipe associations table

parumpum committed:
Formatting

parumpum committed:
Data model fixes for recipe associations

parumpum committed:
Add sub_recipe field to recipes ingredients

parumpum committed:
Use referenced_recipe on ingredient

parumpum committed:
Put linked recipe in its own section with ingredients

parumpum committed:
Change how recipe as ingredient is viewed

parumpum committed:
Add sub-recipe ingredients to shopping list

parumpum committed:
Merge branch 'mealie-recipes:mealie-next' into feat/recipe-links

parumpum committed:
Add recipe reference button text

parumpum committed:
Formatting, fix test

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
Fix dialog setup

parumpum committed:
Show image on sub recipe search dialog

parumpum committed:
Place sub-recipe ingredients in a Section, add select all and none to each section

parumpum committed:
List child recipes on Last Made dialog to optionally update that field

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
Add child recipes to timeline without image and comment

parumpum commented:
Addressed feedback in Discord:

- Added select all / none to Shopping List sections
- Added selectable list of child recipes on "I Made This" dialog
- Images shown when searching for child recipes to help pick the correct item

parumpum committed:
Print option to expand child recipe ingredients

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
Move linked recipe test

felixschndr commented:
Thanks for the great work! Looking forwards to this being merged

felixschndr commented:
@parumpum would you mind fixing the merge conflicts so the MR can be reviewed (@hay-kot)?

Kuchenpirat commented:
we won't merge anything with frontend code before the nuxt 3 upgrade has been done (which will bring new conflicts)

felixschndr commented:
I see, thanks for the clarification (FYI for anyone looking for that PR: https://github.com/mealie-recipes/mealie/pull/5184)

jawilson commented:
Please don't close this PR! I really want this ability.

Kuchenpirat commented:
No worries we are not closing this.
Maintainers are currently just pretty busy with stuff outside of mealie as well as the nuxt 3 upgrade.

tecosaur commented:
As one of the people eagerly awaiting this, now that v3 is out, how's this PR looking?

parumpum commented:
Once I find some free time to update it I’ll get it going!

On Sun, Jul 13, 2025 at 09:58 Timothy ***@***.***> wrote:

> *tecosaur* left a comment (mealie-recipes/mealie#4800)
> <https://github.com/mealie-recipes/mealie/pull/4800#issuecomment-3067166923>
>
> As one of the people eagerly awaiting this, now that v3 is out, how's this
> PR looking?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mealie-recipes/mealie/pull/4800#issuecomment-3067166923>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A4PIWMTU2F7NWAGHM43YWGD3IKF2JAVCNFSM6AAAAABUMT6IPOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANRXGE3DMOJSGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

michael-genson commented:
Hey there! As you might know, we've been working on the upgrade to Nuxt 3 for the past several months, and had to hold on reviewing/merging PRs during that time (especially frontend ones). That work is more-or-less completed, so we've started reviewing PRs again.

This PR has merge conflicts preventing us from fully reviewing/testing it. If you're able to resolve those we can get to work on it.

Appreciate your patience, we know it's been a bit, but now that the Nuxt 3 upgrade is completed we should be able to get things reviewed much quicker than before.

tecosaur commented:
Friendly poke :point_right: @parumpum :slightly_smiling_face:

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum commented:
Just got my dev environment set back up to work on this. Definitely some broken pieces to work on with all the Nuxt 3 changes that went in!

parumpum committed:
Nuxt 3 fixes

parumpum committed:
fix lint errors

parumpum committed:
Remove debug logs and fixed sub-recipe item on ingredients list

parumpum committed:
Fix sub-recipe ingredients scaling in shopping list

parumpum committed:
Utilize Split Button on Ingredient Editor for selecting Food/Recipe/Bulk

parumpum committed:
Ingredient editor cleanup

fleshgolem committed:
fix: add `breaks` option to markdown rendering, to get old linebreak behaviour (#6156)

renovate[bot] committed:
fix(deps): update dependency axios to v1.12.0 [security] (#6158)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

michael-genson committed:
fix: Can't Edit Timeline Events (#6160)

michael-genson committed:
fix: Ingredient Parser Drops Units Sometimes (#6150)

renovate[bot] committed:
fix(deps): update dependency ingredient-parser-nlp to v2.3.0 (#6163)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
docs(auto): Update image tag, for release v3.2.0 (#6164)

Co-authored-by: michael-genson <71845777+michael-genson@users.noreply.github.com>

michael-genson committed:
fix: Explorer Page State Not Working On Hitting Back (#6171)

hay-kot committed:
chore(l10n): New Crowdin updates (#6145)

github-actions[bot] committed:
docs(auto): Update image tag, for release v3.2.1 (#6172)

Co-authored-by: michael-genson <71845777+michael-genson@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency ruff to ^0.13.0 (#6148)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
chore(auto): Update pre-commit hooks (#6174)

Co-authored-by: boc-the-git <3479092+boc-the-git@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update node.js to abcf9c9 (#6138)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency openai to v1.107.3 (#6147)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency pydantic to v2.11.9 (#6159)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency mypy to v1.18.1 (#6161)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency pytest-asyncio to v1.2.0 (#6162)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency mkdocs-material to v9.6.20 (#6179)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency fastapi to v0.116.2 (#6181)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency authlib to v1.6.4 (#6182)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency openai to v1.108.0 (#6185)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

1112luke commented:
I would love this!! Any information on the timeline for this to be merged?

michael-genson committed:
fix: Print Button Does Nothing (#6178)

p0lycarpio committed:
fix: ingredient linker and instructions titles (#6146)

Co-authored-by: Michael Genson <71845777+michael-genson@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency ruff to v0.13.1 (#6191)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

michael-genson committed:
feat: Manually calculate OpenAI Parsing Confidence (#6141)

p0lycarpio committed:
fix: multiple RecipeRating backend calls (#6194)

Co-authored-by: Michael Genson <71845777+michael-genson@users.noreply.github.com>

michael-genson committed:
docs: Add Info Regarding Theme Settings Config (#6198)

michael-genson committed:
feat: Upgraded Ingredient Parsing Workflow (#6151)

hay-kot committed:
chore(l10n): New Crowdin updates (#6176)

hay-kot committed:
chore(l10n): New Crowdin updates (#6213)

github-actions[bot] committed:
chore(l10n): Crowdin locale sync (#6206)

Co-authored-by: GitHub Action <action@github.com>

renovate[bot] committed:
chore(deps): update dependency mypy to v1.18.2 (#6193)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

michael-genson committed:
fix: Actually Fix Token Time (#6215)

lehnerpat committed:
docs: Fix formatting in some community guides (#6223)

github-actions[bot] committed:
chore(auto): Update pre-commit hooks (#6222)

Co-authored-by: boc-the-git <3479092+boc-the-git@users.noreply.github.com>

hay-kot committed:
chore(l10n): New Crowdin updates (#6218)

hay-kot committed:
chore(l10n): New Crowdin updates (#6224)

michael-genson committed:
fix: Remove explicit timeout from OpenAI image API Call (#6227)

hay-kot committed:
chore(l10n): New Crowdin updates (#6225)

michael-genson committed:
feat: Ingredient Parser Enhancements (#6228)

lehnerpat committed:
docs: Update navigation instructions for (admin) settings pages (#6220)

michael-genson committed:
fix: Workflow Issues with Deleting Ingredient In Parser (#6230)

hay-kot committed:
chore(l10n): New Crowdin updates (#6231)

michael-genson committed:
fix: Enabled Using Mealie In iframe If Served Over HTTPS (#6128)

lehnerpat committed:
docs: Improve formatting in 'Automating Backups with n8n' community guide (#6221)

Co-authored-by: Michael Genson <71845777+michael-genson@users.noreply.github.com>

hay-kot committed:
chore(l10n): New Crowdin updates (#6234)

cmintey committed:
fix: Remove constraint on unhashed password being 'LDAP' (#6236)

hay-kot committed:
chore(l10n): New Crowdin updates (#6237)

renovate[bot] committed:
chore(deps): update dependency openai to v1.109.1 (#6196)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency ruff to v0.13.2 (#6239)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency uvicorn to ^0.37.0 (#6200)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency fastapi to ^0.117.0 (#6205)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency coverage to v7.10.7 (#6216)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency lxml to v6.0.2 (#6219)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

michael-genson committed:
fix: Remove Double Cookie Refresh (#6242)

michael-genson committed:
feat: Simplify Default Layout Logic and Add Household.name To Cookbooks API (#6243)

hay-kot committed:
chore(l10n): New Crowdin updates (#6241)

renovate[bot] committed:
chore(deps): update dependency pydantic-settings to v2.11.0 (#6233)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency pyyaml to v6.0.3 (#6245)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency bcrypt to v5 (#6246)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

michael-genson committed:
fix: Misc Issues with Ingredient Parser (#6250)

michael-genson committed:
fix: Make Ingredient Parser Dialog Use Full Space (#6253)

hay-kot committed:
chore(l10n): New Crowdin updates (#6254)

michael-genson committed:
fix: Check for non-hid properties when injetcing SPA meta (#6256)

hay-kot committed:
chore(l10n): New Crowdin updates (#6257)

michael-genson committed:
chore: Upgrade Node and Nuxt (#6240)

michael-genson committed:
chore: Add Stricter Frontend Formatting (#6262)

hay-kot committed:
chore(l10n): New Crowdin updates (#6264)

michael-genson committed:
fix: Stores Not Populating Sometimes (#6266)

renovate[bot] committed:
chore(deps): update dependency beautifulsoup4 to v4.14.0 (#6260)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
docs(auto): Update image tag, for release v3.3.0 (#6267)

Co-authored-by: michael-genson <71845777+michael-genson@users.noreply.github.com>

michael-genson committed:
docs: Update TOKEN_TIME docs to include max (#6279)

renovate[bot] committed:
fix(deps): update dependency fastapi to ^0.118.0 (#6281)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
chore(auto): Update pre-commit hooks (#6282)

Co-authored-by: boc-the-git <3479092+boc-the-git@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency beautifulsoup4 to v4.14.2 (#6283)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

michael-genson committed:
fix: Default to "0" qty when creating ingredients everywhere (#6285)

michael-genson committed:
fix: Restore recipe meta for non-logged-in users (#6286)

renovate[bot] committed:
chore(deps): update dependency tzdata to v2025 (#6287)

renovate[bot] committed:
fix(deps): update dependency openai to v2 (#6294)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency apprise to v1.9.5 (#6290)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update node.js to d367fd3 (#6292)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency pillow-heif to v1.1.1 (#6291)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency mkdocs-material to v9.6.21 (#6293)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Choromanski committed:
fix: Updated pwa orientation to any (#6298)

renovate[bot] committed:
chore(deps): update node.js to 2bb201f (#6295)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency openai to v2.0.1 (#6296)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
docs(auto): Update image tag, for release v3.3.1 (#6300)

Co-authored-by: michael-genson <71845777+michael-genson@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency authlib to v1.6.5 (#6299)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency ruff to v0.13.3 (#6301)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency openai to v2.1.0 (#6302)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

github-actions[bot] committed:
chore(l10n): Crowdin locale sync (#6268)

Co-authored-by: GitHub Action <action@github.com>

hay-kot committed:
chore(l10n): New Crowdin updates (#6273)

hay-kot committed:
chore(l10n): New Crowdin updates (#6309)

hay-kot committed:
chore(l10n): New Crowdin updates (#6313)

hay-kot committed:
chore(l10n): New Crowdin updates (#6317)

github-actions[bot] committed:
chore(l10n): Crowdin locale sync (#6318)

Co-authored-by: GitHub Action <action@github.com>

michael-genson committed:
fix: Re-write Nuxt auth backend and get rid of sidebase auth (#6322)

Coryrichter94 commented:
This is a great feature! Can not wait

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
Fix lint and recipe as ingredient parsing

parumpum committed:
Fix margin on Add split button

parumpum committed:
Deny recursive recipe links

parumpum committed:
Flatten linked recipes for shopping list ingredients

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
Cleanup comments

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

michael-genson committed:
Merge branch 'mealie-next' into feat/recipe-links

michael-genson committed:
fix codegen

michael-genson committed:
cleanup

michael-genson committed:
remove `is_recipe`/`isRecipe` flag

michael-genson committed:
dev:generate

michael-genson commented on mealie/schema/recipe/recipe_ingredient.py:
```diff
@@ -155,15 +156,49 @@ class RecipeIngredientBase(MealieModel):
+        if self.referenced_recipe:
+            self.note = self.referenced_recipe.name
+
+        return self
```
Is this desired behavior? What if we want to add a note to the recipe and have it render? e.g. "Make this thicker since we're using it as a base" or "slice extra thin".

michael-genson reviewed (changes requested):
Looks good so far! I made a few cleanup changes, including removing the `isRecipe` flag. We used to have a bunch of flags on the backend for things like this and it got super messy so we removed them. Similarly to how we handled those, I adjusted the frontend logic to accommodate.

For adding a recipe as an ingredient, instead of adding a dropdown to the "Add" button, would it be better to add it to the existing context menu? This would have the advantage of using an existing ingredient as a recipe instead of adding a new one. Something like "toggle recipe"
<img width="299" height="252" alt="image" src="https://github.com/user-attachments/assets/7a7e2be0-fa97-4408-8cf6-aca446124cbf" />

Another question: when we set a recipe to qty 0, that's signaling to the frontend that we should just not render the quantity (good for things like "teriyaki sauce" which might not have a defined amount). If we add that to the shopping list with 0 qty, will that work, or will nothing for that recipe be added? Might want an edge case for when the qty is 0 to use a shopping list qty of 1.

One more note below as well. I haven't fully finished reviewing this but these were my first impressions :) Love how it looks so far!

parumpum committed:
Fix sub recipe note

parumpum commented:
> Looks good so far! I made a few cleanup changes, including removing the `isRecipe` flag. We used to have a bunch of flags on the backend for things like this and it got super messy so we removed them. Similarly to how we handled those, I adjusted the frontend logic to accommodate.
> 
> For adding a recipe as an ingredient, instead of adding a dropdown to the "Add" button, would it be better to add it to the existing context menu? This would have the advantage of using an existing ingredient as a recipe instead of adding a new one. Something like "toggle recipe" <img alt="image" width="299" height="252" src="https://private-user-images.githubusercontent.com/71845777/505436413-7a7e2be0-fa97-4408-8cf6-aca446124cbf.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE0ODc1MjEsIm5iZiI6MTc2MTQ4NzIyMSwicGF0aCI6Ii83MTg0NTc3Ny81MDU0MzY0MTMtN2E3ZTJiZTAtZmE5Ny00NDA4LThjZjYtYWNhNDQ2MTI0Y2JmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI2VDE0MDAyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyZGZlMjE4ZmJiMjY0MWEwMTJjNTEzYWYyM2I0YzkzMDcyMjQ3M2I4Yzk3ZGZlYTlkOWM1ZDZjZjc0NDEyZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cBFHYvtXDOihdNCJcxTD6q88eiDeXcpdL0HfjynTTPQ">
> 
> Another question: when we set a recipe to qty 0, that's signaling to the frontend that we should just not render the quantity (good for things like "teriyaki sauce" which might not have a defined amount). If we add that to the shopping list with 0 qty, will that work, or will nothing for that recipe be added? Might want an edge case for when the qty is 0 to use a shopping list qty of 1.
> 
> One more note below as well. I haven't fully finished reviewing this but these were my first impressions :) Love how it looks so far!

Setting qty to 0 seems to work as expected. It doesn't show the quantity on the ingredient list, and still allows adding to shopping list as expected.

For the Add button, I feel like expanding the Add Food button is more intuitive than moving this function entirely to the Toggle Section/Insert menu. It does make sense to add a Toggle Recipe option there, so I'll work on that soon.

parumpum committed:
Merge branch 'mealie-next' into feat/recipe-links

parumpum committed:
fix formatting

parumpum committed:
add toggle recipe for ingredient context menu

parumpum committed:
fix formatting

michael-genson committed:
Merge branch 'mealie-next' into feat/recipe-links

michael-genson committed:
improve error handling on has_recursive_recipe_link

michael-genson committed:
expanded tests

michael-genson committed:
add test for deleting a recipe that is linked to

michael-genson committed:
remove unused test

michael-genson committed:
add test for shopping list

michael-genson committed:
add missing backref

michael-genson committed:
remove unused file

michael-genson committed:
rename translation strings

michael-genson committed:
minor cleanup

michael-genson committed:
simplify/refactor ingredient display

michael-genson committed:
regenerate api docs

michael-genson committed:
simplify parse logic

michael-genson committed:
clean up event logic and add new message

michael-genson committed:
update print pref label

michael-genson commented on frontend/components/Domain/Recipe/RecipePrintView.vue:
```diff
@@ -262,32 +262,60 @@ const ingredientSections = computed<IngredientSection[]>(() => {
+        && ingredient.referencedRecipe
+        && ingredient.referencedRecipe.recipeIngredient
+        && ingredient.referencedRecipe.name
+      ) {
+      // Recursively add to the section for this referenced recipe
+        addIngredientsToSections(
+          ingredient.referencedRecipe.recipeIngredient,
+          sections,
+          ingredient.referencedRecipe.name,
+        );
+      }
+      else {
+        const sectionName = title || ingredient.title || "";
+        if (sectionName) {
+          let sec = sections.find(sec => sec.sectionName === sectionName);
+          if (!sec) {
+            sec = { sectionName, ingredients: [] };
+            sections.push(sec);
+          }
+          ingredient.title = sectionName;
+          sec.ingredients.push(ingredient);
+        }
+        else {
+          if (sections.length === 0) {
+            sections.push({
+              sectionName: "",
+              ingredients: [ingredient],
+            });
+          }
+          else {
+            sections[sections.length - 1].ingredients.push(ingredient);
+          }
+        }
+      }
+    });
+  };
```
When printing the linked recipes, if the linked recipe appears in the middle of a section, the section titles become mangled. For instance:
```markdown
Recipe
 - [Section A]
    - ing 1
    - ing 2
    - linked recipe
    - ing 3
```

Becomes:
```markdown
[Section A]
- ing 1
- ing 2
[Linked Recipe]
- linked recipe ing 1
- ...
- ing 3 (from the recipe, not from the linked recipe)
```

Maybe we just keep it simple and remove this functionality from printing?

parumpum replied:
I will remove this functionality. It can be returned with a rework of Sections that better defines which section each ingredient is part of instead of depending on a single ingredient with the "title" property set to define a Section.

Appreciate your assistance getting this through to the finish line.

michael-genson reviewed (changes requested):
Pushed a few more changes/fixes. Just two more things I ran into that I didn't get time to address:
- One regarding printing, see below
- When saving a recursive ingredient, the backend appropriately handles this case and returns an error, but the frontend still looks like it updated successfully. Can we catch these errors on the frontend so the user doesn't think it worked? (we probably don't want to leave edit mode)

parumpum committed:
clean up printing linked recipes and stay in edit mode on error updating recipe

michael-genson committed:
tidy up logic

michael-genson committed:
fix state bug

michael-genson committed:
fix string translation

michael-genson committed:
improve last made style

michael-genson committed:
Merge remote-tracking branch 'upstream/mealie-next' into pr/parumpum/4800

michael-genson committed:
update docs

michael-genson committed:
Merge branch 'mealie-next' into feat/recipe-links

michael-genson reviewed (approved):
LGTM, thanks for sticking with this one! I know it took a while but we got it in there 😅

ambroisie commented:
It's unclear to me how to handle linked recipes amounts. Or rather, I think I understand the intended way to deal with it but find it awkward to read.

Here are a couple of examples:

* "Salted caramel macarons": depends on "Macaron shells". The (unscaled) recipe yield 50 top/bottoms. The salted caramel recipes is scaled for 50 top/bottoms. If I link "1 Macaron shells" that is the correct amount of that recipe to make, but reads weirdly on the page. If I link "50 macaron shells" it links correctly but doesn't reflect the fact that I should 50x the shell recipe.
* "Pain aux raisins": depends on both "Croissant dough" and "Pastry cream". The scales should be 0.5x dough (500 grams from a 1kg yield recipe) and 0.25x cream (200 grams from an 800 gram yield recipe). Once again it reads weirdly to only have "0.5 Croissant dough" on the recipe page, instead of "500 gram croissant dough".

parumpum commented:
> It's unclear to me how to handle linked recipes amounts. Or rather, I think I understand the intended way to deal with it but find it awkward to read.
> 
> Here are a couple of examples:
> 
>     * "Salted caramel macarons": depends on "Macaron shells". The (unscaled) recipe yield 50 top/bottoms. The salted caramel recipes is scaled for 50 top/bottoms. If I link "1 Macaron shells" that is the correct amount of that recipe to make, but reads weirdly on the page. If I link "50 macaron shells" it links correctly but doesn't reflect the fact that I should 50x the shell recipe.
> 
>     * "Pain aux raisins": depends on both "Croissant dough" and "Pastry cream". The scales should be 0.5x dough (500 grams from a 1kg yield recipe) and 0.25x cream (200 grams from an 800 gram yield recipe). Once again it reads weirdly to only have "0.5 Croissant dough" on the recipe page, instead of "500 gram croissant dough".

I think this is more of a servings vs yield issue. Showing "0.5 Croissant dough" on your Pain Aux Raisins recipe indicates that you would scale your croissant dough recipe to 0.5 and make that. If you utilize the shopping list, the ingredients will be scaled to this value.

Likewise for the macarons, you could create or change the recipe to have amounts for 1 macaron shell so that it makes sense to enter 50 macaron shells on the salted caramel macaron recipe. You would be able to set the scale of the macaron shell recipe to the number of shells you ultimately want.

Linking a recipe is simply the quantity of that recipe to make. The current idea behind the feature is that you would be able to go into the main recipe you are ultimately going to make and click on the linked recipe to open a new tab to make that item in the required quantities. One enhancement I could see here would be to pull up the recipe scaled as indicated in the main recipe to save a manual adjustment there.

ambroisie commented:
Yeah I understand, I just find it awkward.

I wish the linked recipe would surface the yield amount and be able to scale that instead.

xF4m3 commented:
Would it be possible to show the ingredients of the linked recipie aswell? Sortes under the linked recipie title or something? I think this would make working with it easier as you do not necessarly have to open the linked recipie.

Also just wanted to say thank you guys for getting this feature implemented in the First place, have been following and waiting for this to happen all this time. Great work, really appreciate this.

ambroisie commented:
> Sorted under the linked recipie title or something

In a foldable menu?

xF4m3 commented:
Doesn't have to be foldabale but that could be one way to do that. In my mind i was just thinking about an expanded static list that is maybe writtin in a smaller or grayer text so it doesn't draw the main attention, but i am probably fine with whatever idea like being collapsible like you mentioned.

MatzeG1t commented:
> Would it be possible to show the ingredients of the linked recipie aswell? Sortes under the linked recipie title or something? I think this would make working with it easier as you do not necessarly have to open the linked recipie.
> 
> Also just wanted to say thank you guys for getting this feature implemented in the First place, have been following and waiting for this to happen all this time. Great work, really appreciate this.

I would suggest to think about the usecase of this. Seeing all ingredients might sound helpful at first, but whats the goal for the user? If its "adding also the ingredient/recipe to the shopping list" a confirmation would make sense at another spot.
Just seeing the ingredients might be only helpful if its a simple ingredient-recipe. If I think about "turkey" or something, the sauce and side dishes are complex as well - only seeing the ingredients in the "turkey recipe" wont help I'd say.

Did the dev team ever thing about adding a feedback function in the tool? This way maybe more user feedback could be sourced. Not just by the mealie-hosts (who check release logs) but rather also the cooks using it - if they are not the same person :)
